### PR TITLE
migration_with_external_tpm: a new test

### PR DESCRIPTION
**Pull request by yanqzhan**
Add a new file for migration with external vtpm device test.
To auto VIRT-296944.

### CODE REVIEW

**Comment by yanqzhan** [comment]
Local run passed:
 (1/1) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.persistent_and_non_p2p: PASS (213.76 s)
(1/1) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.persistent_and_p2p: PASS (213.10 s)
(1/1) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.transient_and_non_p2p: PASS (204.50 s)

with extra settings in avocado-vt/virttest/shared/cfg/base.cfg:
+nfs_mount_src = "/var/lib/avocado/data/avocado-vt/images"
+nfs_mount_dir = "/var/lib/libvirt/migrate"
+nfs_mount_options = "rw"
+export_ip = "*"
+export_options = "rw,no_root_squash"
+export_dir = "/var/lib/avocado/data/avocado-vt/images"
+local_ip = ***
+local_pwd = "******"
+remote_ip = ***
+migrate_source_host = ***
+migrate_source_pwd = ***
+migrate_dest_host = ***
+migrate_dest_pwd = ***

**Comment by dzhengfy** [comment]
@cliping Could you help review?

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            vm.cleanup_serial_console()
+        vm.connect_uri = src_uri
+    if cmd_result:
+        test.fail("Fail to run '%s': %s." % (tpm_cmd, cmd_result))
+
+
+def launch_external_swtpm(params, test, skip_setup=False, on_remote=False):
+    """
+    Launch externally swtpm
+    :param params: dict, test parameters
```
Comment (dzhengfy) [inline libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:76]: Add one blank here.

File: libvirt/tests/cfg/migration_with_vtpm/migration_with_external_tpm.cfg [hunk 2]
```diff
@@ -0,0 +1,50 @@
+- migration_with_vtpm.migration_with_external_tpm:
```
Comment (dzhengfy) [inline libvirt/tests/cfg/migration_with_vtpm/migration_with_external_tpm.cfg:1]: After https://github.com/autotest/tp-libvirt/pull/4895, could you move your case accordingly?

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+from virttest import virsh
+
+from virttest.libvirt_xml import vm_xml
+from virttest.utils_libvirt import libvirt_vmxml
+from virttest.utils_test import libvirt
+
+from provider.migration import base_steps
+
+
+def check_tpm_security_context(params, test, expected_contexts, on_remote=False):
```
Comment (cliping) [inline libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:18]: The same function in tp-libvirt/libvirt/tests/src/migration/migration_with_vtpm/migration_with_shared_tpm.py. Maybe we can create a new file for vtpm in tp-libvirt/provider/ . And then move this function to the new file. @dzhengfy  What's your opinion about this function?

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        cmd_result = remote.run_remote_cmd(cmd, params)
+    else:
+        cmd_result = process.run(cmd, ignore_status=True, shell=True)
+    if cmd_result.exit_status:
+        test.error("Fail to run '%s'." % cmd)
+    if expected_contexts not in cmd_result.stdout_text:
+        test.fail("Fail to find '%s' in '%s'." % (expected_contexts, cmd_result.stdout))
+
+
+def check_vtpm_func(params, vm, test, on_remote=False):
```
Comment (cliping) [inline libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:41]: Same above.

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            process.run('chown qemu:qemu %s' % source_socket, ignore_status=False, shell=True)
+    except Exception as err:
+        if on_remote:
+            remote.run_remote_cmd('pkill swtpm', params)
+        else:
+            process.run("pkill swtpm", shell=True)
+        test.error("{}".format(err))
+
+
+def setup_vtpm(params, test, vm):
```
Comment (cliping) [inline libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:128]: Same above.

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        """
+        tpm_security_contexts = params.get("tpm_security_contexts")
+
+        libvirt.set_vm_disk(vm, params)
+        launch_external_swtpm(params, test)
+        launch_external_swtpm(params, test, skip_setup=False, on_remote=True)
+        setup_vtpm(params, test, vm)
+        check_tpm_security_context(params, test, tpm_security_contexts)
+        check_vtpm_func(params, vm, test)
+        migration_obj.setup_connection
```
Comment (cliping) [inline libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:179]: Missing '()'.

**Review by cliping — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 3 above)
Comment (dzhengfy) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py:18]: Agree

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 1 above)
Comment (yanqzhan) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py]: Done. Thanks!

File: libvirt/tests/cfg/migration_with_vtpm/migration_with_external_tpm.cfg (hunk 2 above)
Comment (yanqzhan) [reply libvirt/tests/cfg/migration_with_vtpm/migration_with_external_tpm.cfg]: Done.

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 6 above)
Comment (yanqzhan) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py]: Removed this line, duplicated with set_vm_disk. Thanks!

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 5 above)
Comment (yanqzhan) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py]: I can optimize them later.

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 4 above)
Comment (yanqzhan) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py]: I can optimize them later.

File: libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py (hunk 3 above)
Comment (yanqzhan) [reply libvirt/tests/src/migration_with_vtpm/migration_with_external_tpm.py]: I can optimize them later.

**Comment by yanqzhan** [comment]
(1/1) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.persistent_and_p2p: PASS (218.65 s)
 (1/2) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.persistent_and_non_p2p: PASS (204.12 s)
 (2/2) type_specific.io-github-autotest-libvirt.migration_with_vtpm.migration_with_external_tpm.transient_and_non_p2p: PASS (210.70 s)

**Review by dzhengfy — APPROVED** [review]

**Review by cliping — APPROVED** [review]
LGTM

**Pull request merged**